### PR TITLE
Updates the code to use refactored mesh functions

### PR DESCRIPTION
**Pull request by bishtgautam**
The code changes include:
1. Use of refactored mesh functions to access various member variables of TDy mesh elements to avoid using `*Offset*`. e.g.

Old approach
```
PetscInt sOffsetFace = subcells->face_offset[subcell_id];

for (iface=0;iface<subcells->num_faces[subcell_id];iface++) {
  PetscInt face_id = subcells->face_ids[sOffsetFace + iface];
  ...
}
```
New approach:
```
PetscInt *face_ids, num_faces;
ierr = TDyMeshGetSubcellFaces(mesh, subcell_id, &face_ids, &num_faces); CHKERRQ(ierr);

for (iface=0;iface<subcells->num_faces[subcell_id];iface++) {
  PetscInt face_id = face_ids[iface];
}
```
2. Conversion of `TDySubcell` and `TDyVector` data to a compressed format.

Note: 
 - There is no need to compress `TDyFace` as it is already in compressed format.
 - `TDyCell` is not converted in the compressed format as it leads to an error that will be fixed in the future.

### CODE REVIEW

File: demo/.gitignore [hunk 1]
```diff
@@ -4,3 +4,7 @@
 *.regression
 *.stdout*
 *.regression.old
+*.txt
```
Comment (jeff-cohere) [inline demo/.gitignore:7]: Haha, I was thinking about sticking these in here myself.

File: src/mpfao/3D/tdympfao3D_steady.c [hunk 2]
```diff
@@ -162,15 +167,19 @@ PetscErrorCode TDyMPFAOComputeSystem_BoundaryVertices_SharedWithInternalVertices
...
       PetscInt iface;
       for (iface=0;iface<subcells->num_faces[subcell_id];iface++) {
 
-        PetscInt face_id = subcells->face_ids[sOffsetFace + iface];
-        PetscInt fOffsetCell = faces->cell_offset[face_id];
-        cell_id_up = faces->cell_ids[fOffsetCell + 0];
-        cell_id_dn = faces->cell_ids[fOffsetCell + 1];
+        PetscInt face_id = face_ids[iface];
+        PetscInt *cell_ids, num_cells;
+        ierr = TDyMeshGetFaceCells(mesh, face_id, &cell_ids, &num_cells); CHKERRQ(ierr);
```
Comment (jeff-cohere) [inline src/mpfao/3D/tdympfao3D_steady.c:179]: Now that I think about this: do all faces have two cells (one on either side)? Or are there cases where a boundary face (say) is only associated with a single cell?

File: src/mpfao/3D/tdympfao3D_core.c [hunk 3]
```diff
@@ -83,22 +83,24 @@ PetscErrorCode TDyComputeGMatrixFor3DMesh(TDy tdy) {
...
-      PetscInt sOffsetFace = subcells->face_offset[subcell_id];
 
-      PetscInt ii,jj;
+      PetscInt *subcell_face_ids, subcell_num_faces;
+      PetscReal *subcell_face_areas;
+      ierr = TDyMeshGetSubcellFaces(mesh, subcell_id, &subcell_face_ids, &subcell_num_faces); CHKERRQ(ierr);
+      ierr = TDyMeshGetSubcellFaceAreas(mesh, subcell_id, &subcell_face_areas, &subcell_num_faces); CHKERRQ(ierr);
 
-      for (ii=0;ii<subcells->num_faces[subcell_id];ii++) {
+      for (PetscInt ii=0;ii<subcells->num_faces[subcell_id];ii++) {
```
Comment (jeff-cohere) [inline src/mpfao/3D/tdympfao3D_core.c:92]: Not a big deal, but could you replace `subcells->num_faces[subcell_id]` here with `subcell_num_faces`? Or is that something different?

Comment (jeff-cohere) [reply src/mpfao/3D/tdympfao3D_core.c:92]: If those things are the same, you could actually replace `subcells->num_faces[subcell_id]` in several places below.

File: src/mpfao/3D/tdympfao3D_core.c [hunk 4]
```diff
@@ -748,31 +759,34 @@ PetscErrorCode ComputeTransmissibilityMatrix_ForNonCornerVertex(TDy tdy,
...
+        PetscInt face_id = subcell_face_ids[iface];
 
-	if (faces->is_internal[face_id] == 0) {
+        PetscInt *face_cell_ids, num_cell_ids;
+        ierr = TDyMeshGetFaceCells(mesh, face_id, &face_cell_ids, &num_cell_ids); CHKERRQ(ierr);
 
-	  // Extract pressure value at the boundary
-	  if (faces->cell_ids[fOffsetCell + 0] >= 0) idxBnd[numBnd] = -faces->cell_ids[fOffsetCell + 1] - 1;
-	  else                                       idxBnd[numBnd] = -faces->cell_ids[fOffsetCell + 0] - 1;
+        if (faces->is_internal[face_id] == 0) {
```
Comment (jeff-cohere) [inline src/mpfao/3D/tdympfao3D_core.c:773]: Can also be spelled `if (!faces->is_internal[face_id]) {`, but that's a matter of taste.

File: src/mesh/tdycoremesh.c [hunk 5]
```diff
@@ -772,7 +772,405 @@ PetscErrorCode SaveMeshConnectivityInfo(TDy tdy) {
...
+///
+/// @param [in] tdy                  A TDy struct
+/// @param [in] num_elements         Number of elements
+/// @param [in] default_offset_size  Default offset size for elements
+/// @param [in] update_offset        Determines if subelement_offset should be updated
+/// @param [inout] subelement_num    Number of subelements for a given element
+/// @param [inout] subelement_offset Offset of subelements for a given element
+/// @param [inout] subelement_id     Subelement ids
+/// @returns 0                       on success, or a non-zero error code on failure
+PetscErrorCode ConvertMeshElementToCompressedFormatIntegerValues(TDy tdy, PetscInt num_element, PetscInt default_offset_size, PetscInt update_offset,
```
Comment (jeff-cohere) [inline src/mesh/tdycoremesh.c:785]: Sorry you had to write these conversion functions, @bishtgautam ! I bet we can avoid doing this in a direct constructor for the mesh. But in the meantime, it's great that things are working.

**Review by jeff-cohere — APPROVED** [review]
To me, the new code definitely reads a bit easier. How do you feel about things after having done all this work, Gautam?

File: src/mpfao/3D/tdympfao3D_steady.c (hunk 2 above)
Comment (bishtgautam) [reply src/mpfao/3D/tdympfao3D_steady.c:179]: All faces have two cells. Boundary faces have cells with negative IDs.

**Comment by bishtgautam** [comment]
@jeff-cohere I too agree that these changes improve code readability.

**Pull request merged**